test/manual/disk_space_monitor_test.cc

```diff
@@ -0,0 +1,196 @@
+    run_system_command(dir, mount_cmd).discard_result().get();
+    auto umount = defer([&] () noexcept {
+        sstring umount_cmd = fmt::format("sudo umount '{}'", mnt_path.native());
+        try {
+            run_system_command(dir, umount_cmd).discard_result().get();
+        } catch (...) {
+            on_fatal_internal_error(testlog, fmt::format("{} failed: {}", umount_cmd, std::current_exception()));
+        }
+    });
+
+    sstring chmod_cmd = fmt::format("sudo chmod 0777 '{}'", mnt_path.native());
+    run_system_command(dir, chmod_cmd).discard_result().get();
+
+    auto space = fs::space(mnt_path);
+    testlog.debug("std::space: capacity={} free={} available={}", space.capacity, space.free, space.available);
+
+    abort_source as;
+    utils::disk_space_monitor::config dsm_config{
+        .sched_group = default_scheduling_group(),
+        .normal_polling_interval = utils::updateable_value<int>(1),
+        .high_polling_interval = utils::updateable_value<int>(1),
+        .polling_interval_threshold = utils::updateable_value<float>(0.5),
+    };
+    utils::disk_space_monitor dsm(as, mnt_path, dsm_config);
+    dsm.start().get();
+    auto stop_dsm = deferred_stop(dsm);
+
+    auto dsm_space = dsm.space();
+    testlog.debug("disk_space_monitor initial space: capacity={} free={} available={} utilization={}", dsm_space.capacity, dsm_space.free, dsm_space.available, dsm.disk_utilization());
+
+    BOOST_REQUIRE_EQUAL(dsm_space.capacity, space.capacity);
+    BOOST_REQUIRE_EQUAL(dsm_space.free, space.free);
+    BOOST_REQUIRE_EQUAL(dsm_space.available, space.available);
+
+    auto foo_name = mnt_path / "foo";
+    write_file(foo_name, image_size * 3 / 4).get();
+
+    sstring syncfs_cmd = fmt::format("sudo sync -f '{}'", foo_name.native());
+    run_system_command(dir, syncfs_cmd).discard_result().get();
+
+    auto st = file_stat(foo_name.native()).get();
+    testlog.debug("foo: allocated_size={} block_size={} blocks={}", st.allocated_size, st.block_size, st.allocated_size / st.block_size);
+
+    std::filesystem::space_info sampled_space;
+    float sampled_disk_utilization = 0;
+    auto sub = dsm.listen([&] (const utils::disk_space_monitor& dsm) -> future<> {
+        sampled_disk_utilization = dsm.disk_utilization();
+        sampled_space = dsm.space();
+        testlog.debug("disk_space_monitor with file: capacity={} free={} available={} utilization={}", sampled_space.capacity, sampled_space.free, sampled_space.available, sampled_disk_utilization);
+        return make_ready_future();
+    });
+
+    auto start_time = lowres_clock::now();
+    auto deadline = start_time + 30s;
+    while (lowres_clock::now() < deadline && sampled_disk_utilization <= 0.5) {
+        sleep(1s).get();
+    }
+
+    BOOST_REQUIRE_GT(sampled_disk_utilization, 0.5);
+    BOOST_REQUIRE_EQUAL(sampled_space.available + st.allocated_size, dsm_space.available);
+}
```

Comment:
test/manual = don't run me

Reply:
I started with putting it in test/boost, but it doesn't have root permissions when running under rootless podman.
I can run the test on a live filesystem, like /tmp, but then it would be flaky.
What do you suggest?

Reply:
Hm. Maybe add an option to supply a function to call instead of fsstat.

It would be pretty silly though. What would we be testing?